src/sql/rewrite/ob_transform_simplify_expr.cpp

```diff
@@ -2517,4 +2526,740 @@ int ObTransformSimplifyExpr::try_remove_ora_decode(ObRawExpr *&expr,
+       * 2. make trans_happened is true 
+      */
+      if (OB_SUCC(ret)) {
+        parent_expr = rewrite_expr;
+        trans_happened = true;
+      } 
+    } /* endif rewrite case_when_expr, when is_valid is true */
+  } /* endif do convert case_when_epxr */
+  return ret;
+}
+
+/***
+ * check one case when can be rewrite or not 
+ * Rewrite Condition:
+ *  1. Let enum_exprs = then_exprs + default_expr
+ *  2. At most one of enum_exprs is in {true, null, non_static}, others are all false (cmp sibling_expr)
+ *    2.1 The result of them cmp sibling_expr are all false, ==> false
+ *    2.2 Only one of them is {true, null, non_static}, record its index
+ *  3. Otherwise, is_valid = false
+*/ 
+int ObTransformSimplifyExpr::check_convert_case_when_by_then_exprs_validity(
+                                              ObRawExpr *&parent_expr,
+                                              ObCaseOpRawExpr *&case_expr,
+                                              ObRawExpr *&sibling_expr,
+                                              const bool case_at_left,
+                                              bool &is_uncalculable,
+                                              int64_t &true_null_uncalc_idx,
+                                              ObRawExpr *&true_null_uncalc_expr,
+                                              ObIArray<ObRawExpr*> &false_exprs,
+                                              bool &is_valid) {
+  int ret = OB_SUCCESS;
+
+  is_uncalculable = false;
+  true_null_uncalc_idx = -1;
+  true_null_uncalc_expr = NULL;
+  false_exprs.reset();
+  is_valid = false;
+
+  if (OB_ISNULL(parent_expr) || OB_ISNULL(case_expr) || OB_ISNULL(sibling_expr)) {
+    ret = OB_ERR_UNEXPECTED;
+    LOG_WARN("unexpected param is NULL", K(ret));
+  } else if (OB_UNLIKELY(case_expr->get_when_expr_size() != case_expr->get_then_expr_size())) {
+    ret = OB_ERR_UNEXPECTED;
+    LOG_WARN("incorrect params of case expr", K(ret));
+  } else if (OB_UNLIKELY(case_expr->get_when_expr_size() < 1)) {
+    ret = OB_ERR_UNEXPECTED;
+    LOG_WARN("incorrect params of case expr", K(ret));
+  } else {
+    const int64_t when_cnt = case_expr->get_when_expr_size();
+    const int64_t enum_value_cnt = when_cnt + 1; // then size + default size
+    ObSEArray<ObRawExpr*, 4> enum_exprs; // includes then_exprs and default_expr
+    // collect then_exprs to enum_exprs
+    for (int64_t i = 0; OB_SUCC(ret) && i < when_cnt; ++i) {
+      ObRawExpr *then_expr = NULL;
+      if (OB_ISNULL(then_expr = case_expr->get_then_param_expr(i))) {
+        ret = OB_ERR_UNEXPECTED;
+        LOG_WARN("param is NULL", K(ret));
+      } else if (OB_FAIL(enum_exprs.push_back(then_expr))) {
+        LOG_WARN("failed to push back enum values");
+      }
+    }
+    // collect default_expr to enum_exprs
+    if (OB_SUCC(ret)) {
+      ObRawExpr *default_expr = NULL;
+      if (OB_ISNULL(default_expr = case_expr->get_default_param_expr())) {
+        ret = OB_ERR_UNEXPECTED;
```

Comment:
just enum_exprs.assign(get_then_param_exprs()) and enum_exprs.push_back() ?
enum_exprs is uncessary, just use get_then_param_exprs() and get_default_param_expr() in the loop.